src/libopensc/card-piv.c

```diff
@@ -506,201 +839,1973 @@ static struct sc_card_driver piv_drv = {
+			sc_log(card->ctx,"KDF hash failed");
+			r = SC_ERROR_INTERNAL;
+			goto err;
+		}
+		kdf_in[3]++;  /* inc the counter */
+		p += cs->kdf_hash_size;
+	}
+
+	/* copy keys used for APDU */
+	memset(&priv->sm_session, 0, sizeof(piv_sm_session_t)); /* clear */
+	priv->sm_session.aes_size = cs->aeskeylen;
+	memcpy(&priv->sm_session.SKcfrm, &aeskeys[cs->aeskeylen * 0], cs->aeskeylen);
+	memcpy(&priv->sm_session.SKmac, &aeskeys[cs->aeskeylen * 1], cs->aeskeylen);
+	memcpy(&priv->sm_session.SKenc, &aeskeys[cs->aeskeylen * 2], cs->aeskeylen);
+	memcpy(&priv->sm_session.SKrmac, &aeskeys[cs->aeskeylen * 3], cs->aeskeylen);
+	sc_mem_clear(&aeskeys, sizeof(aeskeys));
+
+	priv->sm_session.enc_counter[15] = 0x01;
+	priv->sm_session.resp_enc_counter[0] = 0x80;
+	priv->sm_session.resp_enc_counter[15] = 0x01;
+	/* C_MCV is zero */
+	/* R_MCV is zero */
+
+	/*  Step H11 Zeroize Z (and kdf_in which has Z) */
+	if (Z && Zlen) {
+		sc_mem_clear(Z, Zlen);
+		free(Z);
+		Z=NULL;
+		Zlen = 0;
+	}
+	if (kdf_in && kdf_inlen) {
+		sc_mem_clear(kdf_in, kdf_inlen);
+		free(kdf_in);
+		kdf_in = NULL;
+		kdf_inlen = 0;
+	}
+
+	/* Step H12 check AuthCryptogramting our version  */
+	/* Generate CMAC */
+
+	{
+		u8 Check_AuthCryptogram[32];
+		size_t Check_Alen = 0;
+
+		u8 MacData[200];
+		int MacDatalen;
+		memset(MacData, 0, sizeof(MacData));
+
+		p = MacData;
+		memcpy(p, "\x4B\x43\x5f\x31\x5f\x56", 6);
+		p += 6;
+		memcpy(p, IDsicc, cs->IDsicclen);
+		p += cs->IDsicclen;
+		memcpy(p, IDsh, cs->IDshlen);
+		p += cs->IDshlen;
+
+		memcpy(p, Qeh_OS, Qeh_OSlen);
+		p += Qeh_OSlen;
+		MacDatalen = p - MacData;
+
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+		if ((cmac_ctx = CMAC_CTX_new()) == NULL
+				|| CMAC_Init(cmac_ctx, priv->sm_session.SKcfrm, cs->aeskeylen, (*cs->cipher_cbc)(), NULL) != 1
+				|| CMAC_Update(cmac_ctx, MacData, MacDatalen) != 1
+				|| CMAC_Final(cmac_ctx, Check_AuthCryptogram, &Check_Alen) != 1) {
+			r = SC_ERROR_INTERNAL;
+			sc_log(card->ctx,"AES_CMAC failed %d",r);
+			goto err;
+		}
+#else
+		mac = EVP_MAC_fetch(PIV_LIBCTX, "cmac", NULL);
+		cmac_params[cmac_params_n++] = OSSL_PARAM_construct_utf8_string("cipher", cs->cipher_cbc_name, 0);
+
+		cmac_params[cmac_params_n] = OSSL_PARAM_construct_end();
+		if (mac == NULL
+				|| (cmac_ctx = EVP_MAC_CTX_new(mac)) == NULL
+				|| !EVP_MAC_init(cmac_ctx, priv->sm_session.SKcfrm,
+					priv->sm_session.aes_size, cmac_params)
+				|| !EVP_MAC_update( cmac_ctx, MacData, MacDatalen)
+				|| !EVP_MAC_final(cmac_ctx, Check_AuthCryptogram, &Check_Alen, cs->AuthCryptogramlen)) {
+			piv_log_openssl(card->ctx);
+			r = SC_ERROR_INTERNAL;
+			sc_log(card->ctx,"AES_CMAC failed %d",r);
+			goto err;
+		}
+#endif
+
+		if (0 == memcmp(AuthCryptogram, Check_AuthCryptogram, cs->AuthCryptogramlen)) {
+			sc_log(card->ctx,"AuthCryptogram compare");
+			r = 0;
+		} else {
+			sc_log(card->ctx,"AuthCryptogram compare failed");
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
```

Comment:
This error goes unnoticed as the value of `r` is not checked before being used again.

Reply:
Good catch. Fixed by https://github.com/OpenSC/OpenSC/pull/2053/commits/de7bf8b13df4b759f6acafde6459c41ee5cbd244